File: triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        row = len(triangle)
        memo = triangle[row-1].copy()

        for r in range(row-2, -1, -1):
            for c in range(r+1):
                memo[c] = min(memo[c], memo[c+1]) + triangle[r][c]
        
        return memo[0]
                

File: triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int row = triangle.size();
        vector<int> memo = triangle[row-1];

        for (int r = row-2; r >= 0; r--) {
            for (int c = 0; c <= r; c++) {
                memo[c] = min(memo[c], memo[c+1]) + triangle[r][c];
            }
        }

        return memo[0];        
    }
};
